### PR TITLE
Add a bunch of JSDoc comments

### PR DESCRIPTION
**Pull request by neekolas**
## Summary

- Convert `Conversation` type from a union to an interface. This allows for it to appear in the TypeDocs and for the methods to be annotated with comments
- Add a ton of JSDoc comments, and convert others to be picked up by the documentation generator
- Update the README to reflect v8 of the SDK
- Export more types from the root so that they are usable by developers and get picked up in the doc generator

## Notes
- This is the first documentation PR, but I plan to continue the work over the next week.
- To see the generated docs, run `npm run build:docs` and then open `docs/index.html` in your browser (you can just drag the file into the browser to load it).

### CODE REVIEW

File: src/Message.ts [hunk 1]
```diff
@@ -420,3 +396,27 @@ export async function decodeContent(contentBytes: Uint8Array, client: Client) {
 
   return { content, contentType, error }
 }
+
+function conversationReferenceToConversation(
```
Comment (neekolas) [inline src/Message.ts:400]: Moved this out of the class so that it doesn't show in the typedocs. Feels internal and not useful to the public.

**Comment by jhaaaa** [comment]
This is SO EXCITING, @neekolas - thank you for making the time and space to work on this! I'm excited to take a look!

**Comment by neekolas** [comment]
🤦 

I just realized that we have nice CloudFlare preview links for the docs. You can see them at https://nmolnar-documentation-overha.xmtp-js.pages.dev/

File: README.md [hunk 2]
```diff
@@ -137,14 +137,15 @@ const xmtp = await Client.create(wallet)
...
+| keystoreProviders    | `[StaticKeystoreProvider, NetworkKeystoreProvider, KeyGeneratorKeystoreProvider]` | Override the default behaviour of how the Client creates a Keystore with a custom provider. This can be used to get the user's private keys from a different storage mechanism.                                                                                                                                                                                                                                                  |
+| persistConversations | true                                                                              | Maintain a cache of previously seen V2 conversations in the storage provider (defaults to `LocalStorage`).                                                                                                                                                                                                                                                                                                                       |
```
Comment (michaelx11) [inline README.md:145]: mega nit: surround true in backticks for consistency e.g. `true`

File: src/Client.ts [hunk 3]
```diff
@@ -71,33 +66,55 @@ export type XmtpEnv = keyof typeof ApiUrls
...
   env: XmtpEnv
-  // apiUrl can be used to override the default URL for the env
+  /**
+   * apiUrl can be used to override the `env` flag and connect to a
+   * specific endpoint
+   */
   apiUrl: string | undefined
-  // app identifier included with client version header
+  /**
+   * app identifier, which will be included in headers of all API requests
```
Comment (michaelx11) [inline src/Client.ts:80]: any recommended formats? Use their domain name?

**Review by michaelx11 — APPROVED** [review]
Overall LGTM. This isn't related to this PR but wondering if we feel good about updating the security audit statement and linking to [certik](https://www.certik.com/projects/xmtp)?

| xmtp-js has not undergone a formal security audit.

File: src/Client.ts (hunk 3 above)
Comment (neekolas) [reply src/Client.ts:80]: So far, all the apps are just using the app name (ie. `lenster/1.0.7` or `buttrfly/0.0.0`)

File: README.md (hunk 2 above)
Comment (neekolas) [reply README.md:145]: Good call. Updated

File: src/Client.ts (hunk 3 above)
Comment (neekolas) [reply src/Client.ts]: Copied the full description from the README

**Comment by neekolas** [comment]
> This isn't related to this PR but wondering if we feel good about updating the security audit statement and linking to [certik](https://www.certik.com/projects/xmtp)?

We should def include a link. We can still include some context around the scope of the audit.

**Pull request merged**